### ISSUE TITLE
Rule proposal: Disallow switch cases with multiple options

### ISSUE THREAD

**Issue by jfmengels** [sindresorhus/eslint-plugin-unicorn#72]
What I'd wish to disallow is the following:
```js
switch (foo): {
  case 1:
  case 2:
    return bar()
  case 3:
  default:
    baz()
}
```
in favor of

```js
switch (foo): {
  case 1:
    return bar();
  case 2:
    return bar();
  case 3:
    baz();
    break;
  default:
    baz();
}
```

I agree that the former version is better, but there is one case where the second is better, and that is code coverage.

Current coverage tools put markers on every statement / branches, and when the code passes through it, it considers that line/branch as reached and increments the code coverage. Kind of like this:
```js
switch (foo): {
  case 1:
    incrementCodeCoverageForLine(1);
  case 2:
    incrementCodeCoverageForLine(2);
    return bar();
  case 3:
    incrementCodeCoverageForLine(3);
  default:
    incrementCodeCoverageForLine(4);
    baz();
}
```

The problem is how that works when you have switch statements with multiple "conditions". If we take this example (equal to the example that I'd like to disallow), if `foo` equals 1, then we'd pass through both `incrementCodeCoverageForLine(1)` and `incrementCodeCoverageForLine(2)` statements, and the coverage is cheated. If we separated the conditions and enforced having either a return or break statement in each case, then the code coverage is a lot more reliable.

I noticed this when working on [this file](https://github.com/jfmengels/ast-utils/blob/master/lib/contains-identifier.js) where I noticed that (before fixing it) a lot of cases were actually not covered.

If we do this, we should probably mention that this should only be used when 100% code coverage is important, and discourage it otherwise.

**Comment by sindresorhus** [comment]
Sounds like the [`no-fallthrough`](http://eslint.org/docs/rules/no-fallthrough) rule?

**Comment by florianb** [comment]
I hope this issue is not about adding `no-fallthrough` in general -- i can hardly think of a better readable solution to check many occurences (like [here](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/73/commits/f49b4fa3f8faa09c09eb8ece07710a192571cb44)).

Even though it could be Sindre is requiring to change that.

**Comment by jfmengels** [comment]
No, I love `no-fallthrough`, and I highly recommend the use of multiple conditions in a switch statement.
I only want to recommend this when code coverage is important to you, but in general the multiple conditions is your use case.

I totally missed `no-fallthrough` (was searching for `switch` to find similar rules, but the rule title doesn't contain the term unfortunately :sweat_smile:).

That said, `no-fallthrough` allows the case I mentionned, so it's not sufficient. But open an issue to add a new option over there is probably a better solution for the moment.

**Comment by florianb** [comment]
@jfmengels - you are right. In my understanding you're proposing a stricter implementation of no-fallthrough. I think this belongs to eslint. I unfortunately couldn't find an issue discussing this specific topic (although there are some issues treating this as a [feature](https://github.com/eslint/eslint/issues/6588)).

**Comment by jfmengels** [comment]
I actually just opened an issue on istanbul to discuss improving this :https://github.com/istanbuljs/istanbul-lib-instrument/issues/31.

This would be best solved at the coverage reporter level if possible. Let's see how that goes :)

**Comment by sindresorhus** [comment]
Closing as this should be fixed in the code coverage tools instead.
